### ISSUE TITLE
redundant_pattern_match does not trigger on `matches!` macro

### ISSUE THREAD

**Issue by disco07** [rust-lang/rust-clippy#10803]
### Summary

Uses of `matches` macro with `Result` or `Option` in first argument and in the second argument is just here to check if the first argument is `is_ok()` or `is_some()`, do not lint. When I make `cargo expand` on this:
```rust
matches!(Ok::<i32, i32>(10), Ok(_));
```
I have:
```rust
match Ok::<i32, i32>(10) {
        Ok(_) => true,
        _ => false,
    };
```

### Lint Name

redundant_pattern_match

### Reproducer

I tried this code:

```rust
matches!(Ok::<i32, i32>(10), Ok(_));
```

I expected to see this happen:
```rust
Ok::<i32, i32>(10).is_ok()
```
Instead, this happened:
```rust
matches!(Ok::<i32, i32>(10), Ok(_));
```

### Version

```text
rustc 1.71.0-nightly (521f4dae1 2023-05-19)
binary: rustc
commit-hash: 521f4dae1bdf7fe9cf9436ecef9ee7c7442708bf
commit-date: 2023-05-19
host: x86_64-unknown-linux-gnu
release: 1.71.0-nightly
LLVM version: 16.0.4
```

**Comment by disco07** [comment]
@flip1995 I want to work on this, have you some ideas (suggestions) to give me please to resolve this?

**Comment by flip1995** [comment]
This is a bit complicated. The macro expands to a `match` expression. The `redundant_pattern_match` lint i probably has a `span.from_expansion` condition in it, that prevents it from triggering inside of (external) macros. We could loosen that restriction, by checking the `expn_data` if the expansion (directly) comes from the `matches!` macro. I think we have a utility function called `is_expn_of` of something similar for this.

I haven't looked into the code of this lint for quite some time. Let me know, if this explanation already helps you get started, or if you want some more detailed explanation. I'm happy to do that, if necessary :)

**Comment by disco07** [comment]
@flip1995 Thanks I will explore this path
